File: php/gb-authtool.php
#!/usr/bin/env php
<?php
require dirname(__FILE__)."/vendor/Resty.php";

define('CREDS_FILE', "./.gimmebar_user");

/**
 * First, try to load the client id and secret from the config file
 */
define('CONFIG_FILE', __DIR__ . "/config.json");
define('CONFIG_FILE_SAMPLE', __DIR__ . "/config.sample.json");

if (!file_exists(CONFIG_FILE)) {
	echo "No config file exists. Copy ".CONFIG_FILE_SAMPLE." to " .CONFIG_FILE. " and modify\n";
	die();
}

$conf = json_decode(file_get_contents(CONFIG_FILE));

/**
 * initialize the Resty object
 */
$resty = new Resty;
$resty->setBaseUrl($conf->API_BASE_URL);
if (!empty($conf->DEBUG)) {
	$resty->debug(true);
}


/**
 * FIRST STEP: Generate a request token
 */
$resp = $resty->post('auth/token/request', array(
					'type'=>'app',
					'client_id'=>$conf->CLIENT_ID,
					'client_secret'=>$conf->CLIENT_SECRET
				));
$rtok = $resp['body'];

if ($resp['status'] !== 200) {
	echo "Error(s) getting token. Response follows:\n";
	print_r($resp);
	die();
}

$req_token = $rtok->request_token;
echo "Got request token...\n";



/**
 * SECOND STEP: Send the user to the Gimme Bar web site to approve the app
 */

// generate the URL
$auth_url = "{$conf->SITE_BASE_URL}authorize?client_id={$conf->CLIENT_ID}&token={$req_token}&response_type=code";

// prompt to hit ENTER to continue
echo "Hit ENTER to open URL in browser and authenticate on Gimme Bar site. When you're finished, come back here.\n";
$input = fgets(STDIN); // waits for ENTER

// try to open the URL in a browser from the shell
echo "Trying to open <{$auth_url}> ...\nIf this doesn't work, visit the URL manually in a GUI web browser...\n";
shell_exec("open ".escapeshellarg($auth_url));

// tell the user to come back here when he/she is done
echo "Hit ENTER when you've give the app permission to access your account...\n";
$input = fgets(STDIN); // waits for ENTER

/**
 * THIRD STEP: Exchange our now-approved request token for an authorization token
 */
$resp = $resty->post('auth/token/authorization', array(
						'response_type'=>'code',
						'client_id'=>$conf->CLIENT_ID,
						'token'=>$req_token,
					));

if ($resp['status'] !== 200) {
	echo "Error(s) getting authorization. Response follows:\n";
	print_r($resp);
	die();
}

$auth_token = $resp['body']->code;


/**
 * FOURTH STEP: Exchange the authorization token for an access token
 */
$resp = $resty->post('auth/token/access', array(
							'code'=>$auth_token,
							'grant_type'=>'authorization_code',
						));

if ($resp['status'] !== 200) {
	echo "Error(s) getting authorization. Response follows:\n";
	print_r($resp);
	die();
}

$username = $resp['body']->username;
$access_token = $resp['body']->access_token;

echo "Access token retrieved for user {$username}. Testing authentication with Gimme Bar API...\n";

/**
 * Test our credentials
 */
$resty->setCredentials($username, $access_token);
$resp = $resty->get('auth/test');

echo "Response from auth/test:\n";
echo $resp['status'] . "\n";
print_r($resp['headers']);
print_r($resp['body']);

if ($resp['status'] === 200) {
	echo "SUCCESS!!!\n";
	echo "Trying to save credentials to file...\n";
	$fp = file_put_contents(CREDS_FILE, "{$username}:{$access_token}");
	if ($fp) {
		echo "Saved to " . realpath(CREDS_FILE) . "\n";
	} else {
		echo "Failed to write to file " . realpath(CREDS_FILE) . "\n";
	}

} else {
echo "auth fail!!!\n";
}

File: ruby/gb-authtool.rb
require 'yaml'
require 'json'
require 'rest-client'

# Include indifferent_access for some hash niceties
require 'active_support/core_ext/hash/indifferent_access'

# Load in the configuration file
if File.exist? 'config.yml'
  config = YAML.load_file('config.yml').with_indifferent_access
else
  puts "config.yml not found. Please copy the example file and modify."
  exit
end



#
# We start off by obtaining a request token from the Gimme Bar API
#
begin
  response = RestClient.post(
    "#{config[:api_base_url]}auth/token/request",
    :client_id      => config[:client_id],
    :client_secret  => config[:client_secret],
    :type           => 'app'
  )
  response = JSON.parse(response).with_indifferent_access
rescue => e
  puts "Error getting request token."
  puts e.response
  exit
end



#
# Since we now have a request token, we need the account holder to approve it.
#
request_token = response[:request_token]
url = "#{config[:site_base_url]}authorize?client_id=#{config[:client_id]}&token=#{request_token}&response_type=code"

puts <<EOF
Successfully obtained request_token.
Please visit the following URL in your web browser and return here when you're finished.


#{url}


EOF

print "Press Enter to resume..."
gets # Pause execution until they return



#
# Assuming the token has been approved by the account holder, we now exchange
# the request token for an authorization token.
#
begin
  response = RestClient.post(
    "#{config[:api_base_url]}auth/token/authorization",
    :client_id      => config[:client_id],
    :token          => request_token,
    :response_type  => :code
  )
  response = JSON.parse(response).with_indifferent_access
rescue => e
  puts "Error getting authorization token."
  puts e.response
  exit
end



#
# Now that we have the authorization token, we need to exchange that with the
# Gimmebar server for an access token.
#
begin
  response = RestClient.post(
    "#{config[:api_base_url]}auth/token/access",
    :code           => response[:code],
    :grant_type     => :authorization_code
  )
  response = JSON.parse(response).with_indifferent_access
rescue => e
  puts "Error getting access token."
  puts e.response
  exit
end



#
# We've successfully completed the authentication process.
# Let's output the result of our efforts.
#
puts <<EOF
Access token retrieved!

Username:     #{response[:username]}
User ID:      #{response[:user_id]}

Access Token: #{response[:access_token]}
EOF
